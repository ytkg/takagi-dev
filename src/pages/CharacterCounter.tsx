import React, { useState } from 'react';

export default function CharacterCounter() {
  const [input, setInput] = useState('');

  const handleClear = () => {
    setInput('');
  };

  const charCountWithoutNewlines = input.replace(/\n/g, '').length;
  const charCountWithNewlines = input.length;
  const wordCount = input.trim().split(/\s+/).filter(Boolean).length;

  return (
    <div className="p-8">
      <h1 className="text-2xl font-bold mb-4">Character & Word Counter</h1>
      <div className="grid grid-cols-1 gap-4">
        <div>
          <textarea
            className="w-full h-64 p-2 border rounded-md font-mono text-sm bg-gray-50 dark:bg-gray-800 dark:text-white"
            value={input}
            onChange={(e) => setInput(e.target.value)}
            placeholder='Enter text here...'
          />
        </div>
        <div className="flex space-x-4">
            <div className="p-4 border rounded-md bg-gray-50 dark:bg-gray-800">
                <h2 className="text-lg font-semibold">Characters</h2>
                <p className="text-2xl font-bold dark:text-white">{charCountWithNewlines}</p>
            </div>
            <div className="p-4 border rounded-md bg-gray-50 dark:bg-gray-800">
                <h2 className="text-lg font-semibold">Characters (no newlines)</h2>
                <p className="text-2xl font-bold dark:text-white">{charCountWithoutNewlines}</p>
            </div>
            <div className="p-4 border rounded-md bg-gray-50 dark:bg-gray-800">
                <h2 className="text-lg font-semibold">Words</h2>
                <p className="text-2xl font-bold dark:text-white">{wordCount}</p>
            </div>
        </div>
      </div>
      <div className="mt-4 flex space-x-2">
        <button
          onClick={handleClear}
          className="px-4 py-2 bg-gray-500 text-white rounded-md hover:bg-gray-600"
        >
          Clear
        </button>
      </div>
    </div>
  );
}
