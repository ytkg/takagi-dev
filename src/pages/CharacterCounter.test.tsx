import { render, screen, fireEvent } from '@testing-library/react';
import CharacterCounter from './CharacterCounter';
import '@testing-library/jest-dom';

describe('CharacterCounter', () => {
  test('renders the component', () => {
    render(<CharacterCounter />);
    expect(screen.getByText('Character & Word Counter')).toBeInTheDocument();
  });

  test('initial counts are zero', () => {
    render(<CharacterCounter />);
    expect(screen.getByText('Characters')).toBeInTheDocument();
    expect(screen.getByText('Characters (no newlines)')).toBeInTheDocument();
    expect(screen.getByText('Words')).toBeInTheDocument();
    // All three counts should be 0 initially
    const initialCounts = screen.getAllByText('0');
    expect(initialCounts).toHaveLength(3);
  });

  test('updates counts on input without newlines', () => {
    render(<CharacterCounter />);
    const inputArea = screen.getByPlaceholderText('Enter text here...');
    fireEvent.change(inputArea, { target: { value: 'Hello world' } });

    // Characters (with newlines): 'Hello world' -> 11
    // The 'Characters' heading is the parent of the count '11'
    const withNewlinesCount = screen.getByText('Characters').nextElementSibling;
    expect(withNewlinesCount).toHaveTextContent('11');

    // Characters (no newlines): 'Hello world' -> 11
    const withoutNewlinesCount = screen.getByText('Characters (no newlines)').nextElementSibling;
    expect(withoutNewlinesCount).toHaveTextContent('11');

    // Words: 'Hello world' -> 2
    const wordCount = screen.getByText('Words').nextElementSibling;
    expect(wordCount).toHaveTextContent('2');
  });

  test('updates counts on input with newlines', () => {
    render(<CharacterCounter />);
    const inputArea = screen.getByPlaceholderText('Enter text here...');
    fireEvent.change(inputArea, { target: { value: 'Hello\nworld' } });

    // Characters (with newlines): 'Hello\nworld' -> 11
    const withNewlinesCount = screen.getByText('Characters').nextElementSibling;
    expect(withNewlinesCount).toHaveTextContent('11');

    // Characters (no newlines): 'Hello\nworld' -> 10
    const withoutNewlinesCount = screen.getByText('Characters (no newlines)').nextElementSibling;
    expect(withoutNewlinesCount).toHaveTextContent('10');

    // Words: 'Hello\nworld' -> 2
    const wordCount = screen.getByText('Words').nextElementSibling;
    expect(wordCount).toHaveTextContent('2');
  });

  test('clears the input when clear button is clicked', () => {
    render(<CharacterCounter />);
    const inputArea = screen.getByPlaceholderText('Enter text here...');
    fireEvent.change(inputArea, { target: { value: 'Some text' } });

    // Check that counts are updated
    expect(screen.getByText('Characters').nextElementSibling).toHaveTextContent('9');
    expect(screen.getByText('Characters (no newlines)').nextElementSibling).toHaveTextContent('9');
    expect(screen.getByText('Words').nextElementSibling).toHaveTextContent('2');

    const clearButton = screen.getByText('Clear');
    fireEvent.click(clearButton);

    // Check that input is cleared and counts are reset
    expect(inputArea).toHaveValue('');
    expect(screen.getAllByText('0').length).toBe(3);
  });

  test('matches the snapshot', () => {
    const { container } = render(<CharacterCounter />);
    expect(container).toMatchSnapshot();
  });
});
