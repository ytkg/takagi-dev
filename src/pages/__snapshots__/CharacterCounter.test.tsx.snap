// Vitest Snapshot v1, https://vitest.dev/guide/snapshot.html

exports[`CharacterCounter > matches the snapshot 1`] = `
<div>
  <div
    class="p-8"
  >
    <h1
      class="text-2xl font-bold mb-4"
    >
      Character & Word Counter
    </h1>
    <div
      class="grid grid-cols-1 gap-4"
    >
      <div>
        <h2
          class="text-lg font-semibold mb-2"
        >
          Input
        </h2>
        <textarea
          class="w-full h-64 p-2 border rounded-md font-mono text-sm bg-gray-50 dark:bg-gray-800 dark:text-white"
          placeholder="Enter text here..."
        />
      </div>
      <div
        class="flex space-x-4"
      >
        <div
          class="p-4 border rounded-md bg-gray-50 dark:bg-gray-800"
        >
          <h2
            class="text-lg font-semibold"
          >
            Characters
          </h2>
          <p
            class="text-2xl font-bold dark:text-white"
          >
            0
          </p>
        </div>
        <div
          class="p-4 border rounded-md bg-gray-50 dark:bg-gray-800"
        >
          <h2
            class="text-lg font-semibold"
          >
            Characters (no newlines)
          </h2>
          <p
            class="text-2xl font-bold dark:text-white"
          >
            0
          </p>
        </div>
        <div
          class="p-4 border rounded-md bg-gray-50 dark:bg-gray-800"
        >
          <h2
            class="text-lg font-semibold"
          >
            Words
          </h2>
          <p
            class="text-2xl font-bold dark:text-white"
          >
            0
          </p>
        </div>
      </div>
    </div>
    <div
      class="mt-4 flex space-x-2"
    >
      <button
        class="px-4 py-2 bg-gray-500 text-white rounded-md hover:bg-gray-600"
      >
        Clear
      </button>
    </div>
  </div>
</div>
`;
